### PR TITLE
#186: Implement Haplotype data model

### PR DESCRIPTION
**Pull request by reece**
Implements #186

### CODE REVIEW

File: docs/source/appendices/future_plans.rst [hunk 1]
```diff
@@ -265,6 +265,135 @@ Under consideration. See https://github.com/ga4gh/vr-spec/issues/28.
...
+**Notes**
+
+* The term "genotype" has two, related definitions in common use. The
+  narrower definition is a set of alleles observed at a single
+  location and with a ploidy of two, such as a pair of single residue
+  variants on an autosome. The broader, generalized definition is a
+  set of alleles at multiple locations and/or with ploidy other than
+  two.The VR-Spec Genotype entity is based on this broader definition.
+* The term "diplotype" is often used to refer to two haplotypes. The
+  VR-Spec Genotype entity subsumes the conventional definition of
```
Comment (helensch) [inline docs/source/appendices/future_plans.rst:336]: Should the "VR-Spec" be VRS? Also in other places in the genotype description.

File: docs/source/terms_and_model.rst [hunk 2]
```diff
@@ -669,8 +669,9 @@ both current and future.  Variation subclasses are:
 
 **Biological definition**
 
-In biology, variation is often used to mean `genetic variation`_,
-describing the differences observed in DNA among individuals.
+In biology, variation is often used to mean *sequence* variation,
```
Comment (helensch) [inline docs/source/terms_and_model.rst:672]: The lines above this change has. "Variation subclasses are:" and then has "Biological definition"

File: docs/source/terms_and_model.rst [hunk 3]
```diff
@@ -862,7 +863,7 @@ subclasses, but are still treated as variation.
   implementation SHOULD persist the original Text object and respond
   to queries matching the Text object with the new object.
 * Additional Variation subclasses are continually under
-  consideration. Please open a `GitHub issue`_ if you would like to
+  consideration. Please open a `github issue <github-issues>`_ if you would like to
```
Comment (helensch) [inline docs/source/terms_and_model.rst:866]: On readthe docs, the link to github issue, is to https://vr-spec.readthedocs.io/en/186-haplotype/github-issues, and "Page NotF ound"

File: docs/source/appendices/future_plans.rst [hunk 4]
```diff
@@ -265,6 +265,135 @@ Under consideration. See https://github.com/ga4gh/vr-spec/issues/28.
...
+  diplotype. Therefore, the VR-Spec model does not include an explicit
+  entity for diplotypes. See :ref:`this note
+  <genotypes-represent-haplotypes-with-arbitrary-ploidy>` for a
+  discussion.
+* The VR-SPec model makes no assumptions about ploidy of an organism
```
Comment (helensch) [inline docs/source/appendices/future_plans.rst:341]: VR-SPec, has capital P, but from previous comment maybe VRS.

File: docs/source/terms_and_model.rst [hunk 5]
```diff
@@ -876,9 +877,125 @@ subclasses, but are still treated as variation.
...
+     - Limits
+     - Description
+   * - _id
+     - :ref:`CURIE`
+     - 0..1
+     - Variation Id; MUST be unique within document
+   * - type
+     - string
+     - 1..1
+     - Variation type; MUST be "Haplotype" (default)
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:914]: The default refers to the fact that the JSON Schema does specify a default value. However, in this context, I agree that calling out the default is confusing. Removed.

File: docs/source/terms_and_model.rst [hunk 6]
```diff
@@ -876,9 +877,125 @@ subclasses, but are still treated as variation.
...
+
+**Implementation Guidance**
+
+* Haplotypes are an assertion of Alleles known to occur “in cis” or
+  “in phase” with each other.
+* All Alleles in a Haplotype MUST be defined on the same reference
+  sequence.
+* Alleles within a Haplotype MUST not overlap ("overlap" is defined in
+  Interval).
+* The locations of Alleles are independent of upstream “upstream”
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:931]: Yep, upstream upstream is just to emphasize that it's upstream. :-/ (Kidding) Fixed.

File: docs/source/terms_and_model.rst [hunk 7]
```diff
@@ -876,9 +877,125 @@ subclasses, but are still treated as variation.
...
+  Alleles within the Haplotype.
+
+**Sources**
+
+* ISOGG: Haplotype — A haplotype is a combination of alleles (DNA
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:936]: Yes! Good spot. Added.

File: docs/source/terms_and_model.rst [hunk 8]
```diff
@@ -876,9 +877,125 @@ subclasses, but are still treated as variation.
...
+     - Description
+   * - _id
+     - :ref:`CURIE`
+     - 0..1
+     - Variation Id; MUST be unique within document
+   * - type
+     - string
+     - 1..1
+     - Variation type; MUST be "Haplotype" (default)
+   * - members
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:915]: Good catch. Fixed.

File: schema/vr.json [hunk 9]
```diff
@@ -287,6 +329,18 @@
          "type": "object"
       }
    },
+   "enums": {
+      "Completeness": {
+         "default": "UNKNOWN",
+         "description": "Declares the `completeness` of a Haplotype or Genotype definition.  See descriptions in those schemas for additional information. * COMPLETE: The definition is complete. All data are included\n  for the covered assay.\n* PARTIAL: The definition is intentionally incomplete, typically\n  to protect information.  Declaring partial variation allows\n  data submitters to convey that data are missing.\n* UNKNOWN: The definition is not known to be either COMPLETE or\n  PARTIAL.",
```
[the remark this answers is not in the document]
Comment (reece) [reply schema/vr.json:335]: It had been removed from the yaml (which is the schema definition source), but I had not regenerated the json. Fixed.

File: docs/source/terms_and_model.rst [hunk 10]
```diff
@@ -876,9 +877,125 @@ subclasses, but are still treated as variation.
...
+  transmitted together. A haplotype may be one locus, several loci, or
+  an entire chromosome depending on the number of recombination events
+  that have occurred between a given set of loci.
+* SO: haplotype (SO:0001024) — A haplotype is one of a set of
+  coexisting sequence variants of a haplotype block.
+* GENO: Haplotype (GENO:0000871) - A set of two or more sequence
+  alterations on the same chromosomal strand that tend to be
+  transmitted together.
+
+**Examples**
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:947]: Agreed and added both.

File: docs/source/terms_and_model.rst (hunk 2 above)
Comment (reece) [reply docs/source/terms_and_model.rst:672]: Fixed.

File: docs/source/appendices/future_plans.rst (hunk 4 above)
Comment (reece) [reply docs/source/appendices/future_plans.rst:341]: Yep! Done.

File: docs/source/terms_and_model.rst (hunk 3 above)
Comment (reece) [reply docs/source/terms_and_model.rst:866]: Fixed. Good catch!

File: schema/vr.yaml [hunk 11]
```diff
@@ -214,6 +206,32 @@ definitions:
             - $ref: "#/definitions/CURIE"
 
 
+  Haplotype:
+    additionalProperties: false
+    description: >-
+      A set of zero or more Alleles
```
Comment (ahwagner) [inline schema/vr.yaml:212]: See comment above re: SHOULD/MUST and `0..*`/`1..*`. The current definition is `1..*` above, and `zero or more` here.

File: docs/source/terms_and_model.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:866]: Done.

File: docs/source/terms_and_model.rst [hunk 12]
```diff
@@ -876,8 +877,138 @@ subclasses, but are still treated as variation.
...
+* Haplotypes are an assertion of Alleles known to occur “in cis” or
+  “in phase” with each other.
+* All Alleles in a Haplotype MUST be defined on the same reference
+  sequence.
+* Alleles within a Haplotype MUST not overlap ("overlap" is defined in
+  Interval).
+* The locations of Alleles within the Haplotype MUST be interpreted
+  independently.  Alleles that create a net insertion or deletion of
+  sequence MUST NOT change the location of "downstream" Alleles.
+* The `members` attribute is required and SHOULD contain at least one
```
[the remark this answers is not in the document]
Comment (reece) [reply docs/source/terms_and_model.rst:934]: Done. Also changed schema to require at least one element.

File: schema/vr.yaml (hunk 11 above)
Comment (reece) [reply schema/vr.yaml:212]: Done.

**Comment by reece** [comment]
N.B. For mysterious reasons (certainly *something* I did), this branch was marked merged even though it wasn't. Commits that resolved comments no longer longer appeared in this PR.

PR #209 subsumes the above commits, as well as ones that occurred after the phony merge that address comments.

**Pull request merged**